src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php

```diff
@@ -131,71 +140,57 @@ private function extractFromMutator($class, $property)
-            $fromReflectionType = $this->extractFromReflectionType($reflectionType);
+        if ($this->supportsParameterType) {
+            if ($reflectionType = $reflectionParameter->getType()) {
+                $fromReflectionType = $this->extractFromReflectionType($reflectionType);
 
-            if (!$arrayMutator) {
-                return array($fromReflectionType);
+                // HHVM reports not-builtin-"array" as type hints for variadics
+                if ($reflectionType->isBuiltin() || Type::BUILTIN_TYPE_ARRAY !== $fromReflectionType->getBuiltinType()) {
+                    if (!$arrayMutator) {
+                        return array($fromReflectionType);
+                    }
+                    $collectionValueType = $fromReflectionType;
+                }
             }
-
-            $phpType = Type::BUILTIN_TYPE_ARRAY;
-            $collectionKeyType = new Type(Type::BUILTIN_TYPE_INT);
-            $collectionValueType = $fromReflectionType;
-        }
-
-        if ($reflectionParameter->isArray()) {
+        } elseif ($reflectionParameter->isArray()) {
             $phpType = Type::BUILTIN_TYPE_ARRAY;
             $collection = true;
+        } elseif ($this->supportsCallable && $reflectionParameter->isCallable()) {
+            $phpType = Type::BUILTIN_TYPE_CALLABLE;
+        } else {
+            try {
+                if ($typeClass = $reflectionParameter->getClass()) {
+                    $typeClass = $typeClass->name;
+                }
+            } catch (\ReflectionException $e) {
+                // mandatory; extract it from the exception message
+                $typeClass = str_replace(array('Class ', ' does not exist'), '', $e->getMessage());
+            }
+            if ($typeClass) {
+                $phpType = Type::BUILTIN_TYPE_OBJECT;
+            }
         }
 
         if ($arrayMutator) {
             $collection = true;
             $nullable = false;
-            $collectionNullable = $reflectionParameter->allowsNull();
-        } else {
-            $nullable = $reflectionParameter->allowsNull();
-            $collectionNullable = false;
-        }
-
-        if (!isset($collection)) {
-            $collection = false;
-        }
-
-        if (method_exists($reflectionParameter, 'isCallable') && $reflectionParameter->isCallable()) {
-            $phpType = Type::BUILTIN_TYPE_CALLABLE;
-        }
-
-        if ($typeHint = $reflectionParameter->getClass()) {
-            if ($collection) {
-                $phpType = Type::BUILTIN_TYPE_ARRAY;
-                $collectionKeyType = new Type(Type::BUILTIN_TYPE_INT);
-                $collectionValueType = new Type(Type::BUILTIN_TYPE_OBJECT, $collectionNullable, $typeHint->name);
-            } else {
-                $phpType = Type::BUILTIN_TYPE_OBJECT;
-                $typeClass = $typeHint->name;
+            if ($phpType) {
+                $collectionValueType = new Type($phpType, $nullable, $typeClass);
```

Comment:
the behavior for ``$nullable`` does not match the previous code, which was using a ``$collectionNullable`` variable. ``$nullable`` is weird the collection itself is nullable, not whether collection values are